fuse_constraints/test/test_relative_pose_3d_stamped_constraint.cpp

```diff
@@ -0,0 +1,308 @@
+    covariance_blocks.emplace_back(position1->data(), orientation1->data());
+
+    ceres::Covariance::Options cov_options;
+    ceres::Covariance covariance(cov_options);
+    covariance.Compute(covariance_blocks, &problem);
+
+    std::vector<double> cov_pos_pos(position1->size() * position1->size());
+    covariance.GetCovarianceBlock(position1->data(), position1->data(), cov_pos_pos.data());
+
+    std::vector<double> cov_or_or(orientation1->size() * orientation1->size());
+    covariance.GetCovarianceBlock(orientation1->data(), orientation1->data(), cov_or_or.data());
+
+    std::vector<double> cov_pos_or(position1->size() * orientation1->size());
+    covariance.GetCovarianceBlock(position1->data(), orientation1->data(), cov_pos_or.data());
+
+    // Assemble the full covariance from the covariance blocks
+    Eigen::Map<fuse_core::Matrix3d> pos_pos(cov_pos_pos.data());
+    Eigen::Map<fuse_core::Matrix4d> or_or(cov_or_or.data());
+    Eigen::Map<Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> pos_or(cov_pos_or.data());
+
+    fuse_core::Matrix6d actual_covariance;
+    actual_covariance <<
+      pos_pos, pos_or.block<3, 3>(0, 1),
+      pos_or.block<3, 3>(0, 1).transpose(), or_or.block<3, 3>(1, 1);
+
+    fuse_core::Matrix6d expected_covariance = cov_origin;
+    EXPECT_TRUE(expected_covariance.isApprox(actual_covariance, 1.0e-9));
+  }
+
+  // Compute the marginal covariance for pose2
+  {
+    std::vector<std::pair<const double*, const double*> > covariance_blocks;
+    covariance_blocks.emplace_back(position2->data(), position2->data());
+    covariance_blocks.emplace_back(position2->data(), orientation2->data());
+    covariance_blocks.emplace_back(orientation2->data(), orientation2->data());
+
+    ceres::Covariance::Options cov_options;
+    ceres::Covariance covariance(cov_options);
+    covariance.Compute(covariance_blocks, &problem);
+
+    std::vector<double> cov_pos_pos(position2->size() * position2->size());
+    covariance.GetCovarianceBlock(position2->data(), position2->data(), cov_pos_pos.data());
+
+    std::vector<double> cov_or_or(orientation2->size() * orientation2->size());
+    covariance.GetCovarianceBlock(orientation2->data(), orientation2->data(), cov_or_or.data());
+
+    std::vector<double> cov_pos_or(position2->size() * orientation2->size());
+    covariance.GetCovarianceBlock(position2->data(), orientation2->data(), cov_pos_or.data());
+
+    // Assemble the full covariance from the covariance blocks
+    Eigen::Map<fuse_core::Matrix3d> pos_pos(cov_pos_pos.data());
+    Eigen::Map<fuse_core::Matrix4d> or_or(cov_or_or.data());
+    Eigen::Map<Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> pos_or(cov_pos_or.data());
+
+    fuse_core::Matrix6d actual_covariance;
+    actual_covariance <<
+      pos_pos, pos_or.block<3, 3>(0, 1),
+      pos_or.block<3, 3>(0, 1).transpose(), or_or.block<3, 3>(1, 1);
+
+    fuse_core::Matrix6d expected_covariance;
+    expected_covariance <<
```

Comment:
Where did you get the expected covariance from? I would have naively expected the rotation covariance to be near 2.0.

Reply:
So, funny you should ask. These terms are a result of the `2.0` term in the orientation residuals. If I remove them, the covariance is 2, just as you expect.

Furthermore, if I _add_ the 2.0 term to the orientation cost functor, it breaks the prior covariance check here:

https://github.com/locusrobotics/fuse/pull/27/files#diff-49f598a9a090323e4f949817796747c2R258

Reply:
Does that mean these 2.0 terms are a bad idea? I feel like they should exist everywhere, or nowhere. I suppose I'm going to have to read that 95 page documents. :(

Also, have you played with that `GetCovarianceBlockInTangentSpace()` function you showed me at some point? I'm wondering if that is part of the answer as well.

Reply:
No, I think they're probably fine, but it means that we'll either have to (a) see how the tangent space covariance works out, or (b) understand what the ramifications are for the covariance as-is.